independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java

```diff
@@ -96,28 +96,35 @@ public BeanGenerator(AnnotationLiteralProcessor annotationLiterals, Predicate<Do
     /**
      *
      * @param bean
+     * @param existingClasses
+     * @param beanToGeneratedName
      * @return a collection of resources
      */
-    Collection<Resource> generate(BeanInfo bean, ReflectionRegistration reflectionRegistration) {
+    Collection<Resource> generate(BeanInfo bean, ReflectionRegistration reflectionRegistration, Set<String> existingClasses,
+            Map<BeanInfo, String> beanToGeneratedName) {
         if (bean.getTarget().isPresent()) {
             AnnotationTarget target = bean.getTarget().get();
             switch (target.kind()) {
                 case CLASS:
-                    return generateClassBean(bean, target.asClass(), reflectionRegistration);
+                    return generateClassBean(bean, target.asClass(), reflectionRegistration, existingClasses,
+                            beanToGeneratedName);
                 case METHOD:
-                    return generateProducerMethodBean(bean, target.asMethod(), reflectionRegistration);
+                    return generateProducerMethodBean(bean, target.asMethod(), reflectionRegistration, existingClasses,
+                            beanToGeneratedName);
                 case FIELD:
-                    return generateProducerFieldBean(bean, target.asField(), reflectionRegistration);
+                    return generateProducerFieldBean(bean, target.asField(), reflectionRegistration, existingClasses,
+                            beanToGeneratedName);
                 default:
                     throw new IllegalArgumentException("Unsupported bean type");
             }
         } else {
             // Synthetic beans
-            return generateSyntheticBean(bean, reflectionRegistration);
+            return generateSyntheticBean(bean, reflectionRegistration, existingClasses, beanToGeneratedName);
```

Comment:
We can't do this for synthetic beans because they may depend on application classes, either any of the attribute or even the existence could be conditioned. An example - for every class `X` annotated with `@Foo` generate a synthetic bean with bean types `X` and all interfaces it implements.

Reply:
If it is loaded by the base CL then then the bean is immutable anyway, even if you generate it again the changes to not take effect.

Reply:
But the CL for non-app classes should be discarded, or?

Reply:
the CL for non app beans is not discarded, only for app beans

Reply:
I'm pretty sure we already run into this kind of problem. But I cannot find the relevant issue or code... AFAIR we pretended that the synthetic bean is an app class.

Reply:
> the CL for non app beans is not discarded, only for app beans

Yep, that was I typo I spotted too late ;-).

Reply:
I've created https://github.com/quarkusio/quarkus/issues/8877.